### ISSUE TITLE
"datastore: key not found" / "failed to lookup index for mh"

### ISSUE THREAD

**Issue by RobQuistNL** [filecoin-project/boost#499]
### Checklist

- [X] This is **not** a security-related bug/issue. If it is, please follow please follow the [security policy](https://github.com/filecoin-project/lotus/security/policy).
- [X] This is **not** a question or a support request. If you have any lotus related questions, please ask in the [lotus forum](https://github.com/filecoin-project/lotus/discussions).
- [X] This is **not** a new feature request. If it is, please file a [feature request](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Ffeature&template=feature_request.yml) instead.
- [X] This is **not** an enhancement request. If it is, please file a [improvement suggestion](https://github.com/filecoin-project/lotus/issues/new?assignees=&labels=need%2Ftriage%2Ckind%2Fenhancement&template=enhancement.yml) instead.
- [X] I **have** searched on the [issue tracker](https://github.com/filecoin-project/lotus/issues) and the [lotus forum](https://github.com/filecoin-project/lotus/discussions), and there is no existing related issue or discussion.
- [X] I am running the [`Latest release`](https://github.com/filecoin-project/lotus/releases), or the most recent RC(release canadiate) for the upcoming release or the dev branch(master), or have an issue updating to any of these.
- [ ] I did not make any code changes to lotus.

### Lotus component

- [ ] lotus daemon - chain sync
- [ ] lotus miner - mining and block production
- [ ] lotus miner/worker - sealing
- [ ] lotus miner - proving(WindowPoSt)
- [X] lotus miner/market - storage deal
- [X] lotus miner/market - retrieval deal
- [X] lotus miner/market - data transfer
- [ ] lotus client
- [ ] lotus JSON-RPC API
- [ ] lotus message management (mpool)
- [ ] Other

### Lotus Version

```text
1.15.1
```

### Describe the Bug

I have a host that
1) has data sealed & proving
2) has fast retrieval enabled for this deal
3) has the data unsealed available
4) has had many restarts of the markets actor because it freezes / breaks

But when trying to retrieve the data from the host, the retrieve command returns the following;

```
$ lotus client retrieve --provider f01392893 --car Qmf1ykhUo63qB5dJ8KRyeths9MZfyxpVdT5xi1moLKefz7 ~/testingRetrieval.car
ERROR: offer error: retrieval query offer errored: failed to fetch piece to retrieve from: getting pieces for cid Qmf1ykhUo63qB5dJ8KRyeths9MZfyxpVdT5xi1moLKefz7: getting pieces containing block Qmf1ykhUo63qB5dJ8KRyeths9MZfyxpVdT5xi1moLKefz7: failed to lookup index for mh 1220f7ce2d20772b959c1071868e9495712f12785b1690ee88752af120dd49338190, err: datastore: key not found
```

Is the markets actor broken? Does it not know where to find the data? How can we repair it?

### Logging Information

```text
-
```

### Repo Steps

-

**Comment by brendalee** [comment]
thanks for reporting! tagging @dirkmc @nonsense and @LexLuthr for awareness. 

Tagging a dagstore related PR which could be helpful: https://github.com/filecoin-project/boost/pull/494#

**Comment by RobQuistNL** [comment]
I have just ran the entire re-indexing (which took almost 2 days) but I still can not retrieve some deals that are 100% active on chain.

**Comment by RobQuistNL** [comment]
```
lotus client ls --provider=f01392893 QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ
ERROR: retrieve: offer error: retrieval query offer errored: failed to fetch piece to retrieve from: getting pieces for cid QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ: getting pieces containing block QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ: failed to lookup index for mh 122071a7aeb3eb45f3a054cd64eb683291868550e870c1cfab1b9a306c55a2ce6431, err: datastore: key not found
```

```
$ lotus client ls --provider=f01392893 --pieceCid baga6ea4seaqnho54pbhiwiz5iltkb3lmexcw4bwejdy7gpqx35cdnzamtlbt4ba QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ
ERROR: retrieve: offer error: retrieval query offer errored: failed to fetch piece to retrieve from: getting pieces for cid QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ: getting pieces containing block QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ: failed to lookup index for mh 122071a7aeb3eb45f3a054cd64eb683291868550e870c1cfab1b9a306c55a2ce6431, err: datastore: key not found
```

but the piecestore shows it:
```
master@daemon:~$ lotus-miner pieces cid-info QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ
Info for:  QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ
PieceCid                                                          Offset  Size
baga6ea4seaqnho54pbhiwiz5iltkb3lmexcw4bwejdy7gpqx35cdnzamtlbt4ba  0       0
master@daemon:~$ lotus-miner pieces piece-info baga6ea4seaqnho54pbhiwiz5iltkb3lmexcw4bwejdy7gpqx35cdnzamtlbt4ba
Piece:  baga6ea4seaqnho54pbhiwiz5iltkb3lmexcw4bwejdy7gpqx35cdnzamtlbt4ba
Deals:
DealID   SectorID  Length       Offset
4785546  152358    34359738368  0
```

```
master@daemon:/mnt/md0/dagstore$ lotus-miner --call-on-markets dagstore initialize-all --concurrency 5 --include-sealed
(4/9205) baga6ea4seaqkuhx7eqsfvkldaefzbc3y5xrrdjxcgr7tu7u4ak2cnyyej4zegny SUCCESS
```

```
$ lotus client ls --provider=f01392893 baga6ea4seaqkuhx7eqsfvkldaefzbc3y5xrrdjxcgr7tu7u4ak2cnyyej4zegny
ERROR: retrieve: offer error: retrieval query offer errored: failed to fetch piece to retrieve from: getting pieces for cid baga6ea4seaqkuhx7eqsfvkldaefzbc3y5xrrdjxcgr7tu7u4ak2cnyyej4zegny: getting pieces containing block baga6ea4seaqkuhx7eqsfvkldaefzbc3y5xrrdjxcgr7tu7u4ak2cnyyej4zegny: failed to lookup index for mh 922020aa1eff24245aa963010b908b78ede311a6e2347f3a7e9c02b426e3044f324337, err: datastore: key not found
```

**Comment by dirkmc** [comment]
@[RobQuistNL](https://github.com/RobQuistNL) I'm not sure what's going wrong here. As part of the DAG store initialization process the DAG store:
- Gets the data for the piece from the sealer
- Extracts the list of multihashes in the piece and adds them to two indexes:
  1. An index from `<piece cid> => <all multihashes in the piece>`
  2. An index from `<multihash> => <cids of all pieces containing the multihash>`

The output of initialize-all that you pasted shows that the indexing completed correctly for the piece.

When the client tries to retrieve a file using the payload CID:
1. The client sends the payload CID
2. The provider looks up the multihash of the payload CID in the index of `<multihash> => <piece cids>`
If it can't find the multihash in this index it outputs the error you're seeing: `failed to lookup index for mh`

Something I noticed that seems odd in the output that you pasted:
```
$ lotus-miner pieces cid-info QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ
Info for:  QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ
PieceCid                                                          Offset  Size
baga6ea4seaqnho54pbhiwiz5iltkb3lmexcw4bwejdy7gpqx35cdnzamtlbt4ba  0       0
```

The piece should not have a size of zero. I wonder if there's a problem with that particular CAR file.

Are you seeing this same error when attempting to retrieve other pieces?

**Comment by RobQuistNL** [comment]
Yes, this error happens with a lot of my deals. They have mostly been generated like so;

```
ipfs add -r --nocopy --pin=false folder > folder.out
```

then I check the last line of the .out (which should have the dag root hash)
I export that;
```
ipfs dag export -p=false dag_root > folder.car
```

and then that one is being used to make deals & seal.

I know it worked before - I was able to list the contents of that file using the `ls` commands and it all worked fine.

I have ~12k car files which I did that way, here's another;

```
$ lotus-miner pieces cid-info QmP2z7DdgLLeqBMxvcB3VJjr1wdCtLwzJZFuBw8jh5fZg5
Info for:  QmP2z7DdgLLeqBMxvcB3VJjr1wdCtLwzJZFuBw8jh5fZg5
PieceCid                                                          Offset  Size
baga6ea4seaqh77kzhlxsmetkntmnclnoxyz2arxs6bilnalydilecmfty7yx6mq  0       0
```

**Comment by dirkmc** [comment]
@RobQuistNL I'm going to do some testing today and try to reproduce this issue

**Comment by RobQuistNL** [comment]
Here's a list of providerID's and retrievals that failed because of this bug;


| Provider ID | Failure count by blocks missing |
| ------------- | ------------- |
| f010617 | 179 |
| f01199430 | 53 |
| f01199442 | 76 |
| f01201327 | 57 |
| f01207045 | 55 |
| f01208862 | 55 |
| f01240 | 290 |
| f01278 | 163 |
| f0127896 | 1 |
| f014768 | 111 |
| f020378 | 46 |
| f022352 | 79 |
| f0230200 | 1 |
| f02620 | 4 |
| f030379 | 1 |
| f0400920 | 2 |
| f0504054 | 5 |
| f0694396 | 2 |
| f0707721 | 257 |
| f08399 | 315 |

This is from running my evergreen bot for about 1 hour

**Comment by dirkmc** [comment]
I was able to repro on our miner, I'm debugging now

**Comment by dirkmc** [comment]
So in my case the problem was actually that the deal failed, so it wasn't indexed correctly.

What happens if you do
```
lotus-miner dagstore initialize-shard <piece cid>
```

Note that in this example you provided it may not be working because you're trying to retrieve by piece cid (instead of payload cid):
```
$ lotus client ls --provider=f01392893 baga6ea4seaqkuhx7eqsfvkldaefzbc3y5xrrdjxcgr7tu7u4ak2cnyyej4zegny
```
(initialize-all outputs piece cids, not payload cids)

**Comment by RobQuistNL** [comment]
```
master@daemon:~$ lotus-miner dagstore initialize-shard baga6ea4seaqh77kzhlxsmetkntmnclnoxyz2arxs6bilnalydilecmfty7yx6mq
ERROR: failed to get shard info: shard not found

master@daemon:~$ lotus-miner dagstore initialize-shard QmP2z7DdgLLeqBMxvcB3VJjr1wdCtLwzJZFuBw8jh5fZg5
ERROR: failed to get shard info: shard not found

master@daemon:~$ lotus-miner dagstore initialize-shard bafybeiakkuvje3ztt7rd2u6fb743clyyk44crjd2bpuxrmgmhz7a6uxori
ERROR: failed to get shard info: shard not found
```

This is carfile # 23:
root dag: `QmP2z7DdgLLeqBMxvcB3VJjr1wdCtLwzJZFuBw8jh5fZg5`
piece cid: `baga6ea4seaqh77kzhlxsmetkntmnclnoxyz2arxs6bilnalydilecmfty7yx6mq`
base32 datacid: `bafybeiakkuvje3ztt7rd2u6fb743clyyk44crjd2bpuxrmgmhz7a6uxori`

deal on f01392893:
dealCID: bafyreiggp6lifu4prhzld6ejduddidobeie7i5dv7ug2ueldmjziqastki
dealID: 3095799
Sector ID: 118718


Deal Status:
```
Dec 27 23:39:48  true      bafyreiggp6lifu4prhzld6ejduddidobeie7i5dv7ug2ueldmjziqastki  3095799  StorageDealError           f3w436qzcbnzpngwwkslethcbsnt7rjdslf25mk7ufoo27gmf4uxt3pa3yn5zilzaqwqduew7pqmu5v2nypuiq  32GiB   0 FIL              1443479                                                                                                                                  error awaiting deal pre-commit: failed to set up called handler: called check error (h: 1410528): failed to look up deal on chain: deal 3095799 not found - deal may not have completed sealing before deal proposal start epoch, or deal may have been slashed
```

**Comment by RobQuistNL** [comment]
But yeah, if the index provider only receives "active" deals from the deal subsystem, then thats not going to work because that 2-year-old bug is snowballing through to the indexing subsystem.

**Comment by RobQuistNL** [comment]
More related bugs: 
https://github.com/filecoin-project/lotus/issues/5126 
https://github.com/filecoin-project/lotus/issues/7385

**Comment by RobQuistNL** [comment]
```
$ lotus state get-deal 3095799
{
  "Proposal": {
    "PieceCID": {
      "/": "baga6ea4seaqh77kzhlxsmetkntmnclnoxyz2arxs6bilnalydilecmfty7yx6mq"
    },
    "PieceSize": 34359738368,
    "VerifiedDeal": true,
    "Client": "f01615359",
    "Provider": "f01392893",
    "Label": "QmP2z7DdgLLeqBMxvcB3VJjr1wdCtLwzJZFuBw8jh5fZg5",
    "StartEpoch": 1434439,
    "EndEpoch": 2877918,
    "StoragePricePerEpoch": "0",
    "ProviderCollateral": "4975603048002379",
    "ClientCollateral": "0"
  },
  "State": {
    "SectorStartEpoch": 1412359,
    "LastUpdatedEpoch": 1791159,
    "SlashEpoch": -1
  }
}
```

**Comment by dirkmc** [comment]
I'm not so familiar with this part of the code, I'm looking through how it works as we go.

So it seems like there are a couple of stages that a deal goes through with the DAG store:
1. When the deal has been handed off to the sealer, it is `Registered`
   This means the DAG store knows about the deal (knows its piece CID)
2. In the background, the deal is `Initialized`
  This means the DAG store gets the deal data from the sealer, and builds and index of its cids

When you run `initialize-shard` with the deal's piece CID, it will only work if the DAG store already knows about the deal: ie the deal's piece was `Registered`. The reason you're getting `shard not found` is because the deal wasn't correctly registered with the DAG store.

So I think we need a command that will allow you to manually register a deal with the DAG store. I will add this as a ticket to Boost and close out this ticket.

**Comment by RobQuistNL** [comment]
I initially opened this issue in the lotus project - shouldn't the root cause be fixed there instead of having a patch / workaround here?

**Comment by dirkmc** [comment]
The DAG store is part of the markets subsystem, which is now being replaced by Boost. So the fix should be in Boost.

**Comment by dirkmc** [comment]
Closing in favour of https://github.com/filecoin-project/boost/issues/509

**Comment by RobQuistNL** [comment]
I mean if there's a way that we can patch it with a command like this, maybe we can at least restore our dagstore to a normal state using that.

The underlaying issue still seems to exist though. Here's a recent deal thats active and probably unretrievable;

```json
master@daemon:~$ lotus state get-deal 5816931
{
  "Proposal": {
    "PieceCID": {
      "/": "baga6ea4seaqndlukp7vphibmpxn32rjhjqt6yhw7oo3fs2ktdzkys4zmk437iiq"
    },
    "PieceSize": 34359738368,
    "VerifiedDeal": true,
    "Client": "f01817237",
    "Provider": "f01392893",
    "Label": "uAXASILL83AGD7C185VjkO5CHUFTA0gKNLpeEgCGdY3fNRbP9",
    "StartEpoch": 1813627,
    "EndEpoch": 3237918,
    "StoragePricePerEpoch": "0",
    "ProviderCollateral": "5911184172583188",
    "ClientCollateral": "0"
  },
  "State": {
    "SectorStartEpoch": 1773590,
    "LastUpdatedEpoch": -1,
    "SlashEpoch": -1
  }
}
```

```
master@daemon:~$ cat ~/storagedeals.txt | grep "5816931"
May  2 15:34:18  true   bafyreigzonrfgfx5vessznhf6nzrtw3dirzkja2milexpk7ualuwcqnkyu  5816931  StorageDealError              f1efk7kg5hgmd2r6yr4kyro24pou5hd64ua7x2iii                                               32GiB   0 FIL             1424291                                                                                                                                 error awaiting deal pre-commit: failed to set up called handler: called check error (h: 1768288): failed to look up deal on chain: deal 5816931 not found - deal may not have completed sealing before deal proposal start epoch, or deal may have been slashed
```

```
master@daemon:~$ lotus-miner pieces piece-info baga6ea4seaqndlukp7vphibmpxn32rjhjqt6yhw7oo3fs2ktdzkys4zmk437iiq
Piece:  baga6ea4seaqndlukp7vphibmpxn32rjhjqt6yhw7oo3fs2ktdzkys4zmk437iiq
Deals:
DealID   SectorID  Length       Offset
5816931  161954    34359738368  0
```

```
master@daemon:~$ lotus-miner pieces cid-info uAXASILL83AGD7C185VjkO5CHUFTA0gKNLpeEgCGdY3fNRbP9
Info for:  bafybeifs7toada7mfv6okwhehoiioucuydjafdjos6ciaim5mn342rnt7u
PieceCid                                                          Offset  Size
baga6ea4seaqndlukp7vphibmpxn32rjhjqt6yhw7oo3fs2ktdzkys4zmk437iiq  0       0
```

**Comment by RobQuistNL** [comment]
I can provide a list of dealCID's / dealID's that are active on chain but are marked as failed by doing some simple CLI stuff with the tools available now - is there a way I can allow these to work in my "legacy" lotus markets setup? Or do I need to wait on #509 ?

**Comment by RobQuistNL** [comment]
I can safely say about 70% of my deals are currently unretrievable. Looking at other providers, I feel like they have similar values.

**Comment by RobQuistNL** [comment]
@LexLuthr

**Comment by kernelogic** [comment]
Tried several different miners and different CIDs, all same error.
![image](https://user-images.githubusercontent.com/3711825/178657248-c5553d47-5307-4b95-b840-2313fab64928.png)

**Comment by LexLuthr** [comment]
@kernelogic This is fixed in boost if you want to give it a try. Otherwise, it should be available in upcoming lotus stable release.
We need to to add the missing pieceCID in dagstore and retrievals should work. Currently, deals containing these files have not been indexed correctly by the dagstore.

**Comment by RobQuistNL** [comment]
Now that 1.17.0 is finally stable;

Before:
```
$ lotus client ls --provider=f01392893 --pieceCid baga6ea4seaqnho54pbhiwiz5iltkb3lmexcw4bwejdy7gpqx35cdnzamtlbt4ba QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ
ERROR: retrieve: offer error: retrieval query offer errored: failed to fetch piece to retrieve from: getting pieces for cid QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ: getting pieces containing block QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ: failed to lookup index for mh 122071a7aeb3eb45f3a054cd64eb683291868550e870c1cfab1b9a306c55a2ce6431, err: datastore: key not found
```

Then ran:
```
$ lotus-miner dagstore register-shard baga6ea4seaqnho54pbhiwiz5iltkb3lmexcw4bwejdy7gpqx35cdnzamtlbt4ba
Registered shard baga6ea4seaqnho54pbhiwiz5iltkb3lmexcw4bwejdy7gpqx35cdnzamtlbt4ba
```

Then output was:
```
$ lotus client ls --provider=f01392893 --pieceCid baga6ea4seaqnho54pbhiwiz5iltkb3lmexcw4bwejdy7gpqx35cdnzamtlbt4ba QmVzK5WbFrVta19J3nbiMTE46XsBFCWSYFKzNHsEyo9NXJ
Recv 0 B, Paid 0 FIL, Open (New), 1ms [1659702364201620552|0]
Recv 0 B, Paid 0 FIL, DealProposed (WaitForAcceptance), 3ms [1659702364201620552|0]
Recv 0 B, Paid 0 FIL, DealAccepted (Accepted), 22ms [1659702364201620552|0]
Recv 0 B, Paid 0 FIL, PaymentChannelSkip (Ongoing), 22ms [1659702364201620552|0]
Recv 59 B, Paid 0 FIL, BlocksReceived (Ongoing), 143ms [1659702364201620552|59]
Recv 59 B, Paid 0 FIL, AllBlocksReceived (BlocksComplete), 146ms [1659702364201620552|59]
Recv 59 B, Paid 0 FIL, Complete (CheckComplete), 146ms [1659702364201620552|59]
Recv 59 B, Paid 0 FIL, CompleteVerified (FinalizingBlockstore), 147ms [1659702364201620552|59]
Recv 59 B, Paid 0 FIL, BlockstoreFinalized (Completed), 147ms [1659702364201620552|59]

QmUwQk2EhtpqRPop7cS372Vc4VBGSsex3yRMFAG3xXHHc8 PROJ14724	17609480856
```

So this fixes the issue. Now to automate it :)

**Comment by beck-8** [comment]
https://github.com/filecoin-project/dagstore/issues/160
